Add Level 1 routine sdot (#260)
stdlib-js/stdlib: lib/node_modules/@stdlib/blas/base/sdot/README.md

<!--

@license Apache-2.0

> Copyright (c) 2018 The Stdlib Authors.

Reviewer: We'll want to change the copyright year here and elsewhere from 2018 to 2019.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# sdot

> Calculate the dot product of `x` and `y`.

<section class="usage">

## Usage

```javascript
var sdot = require( '@stdlib/blas/base/sdot' );
```

#### sdot( N, x, strideX, y, strideY )

> Takes the dot product of vectors `x` and `y`.

Reviewer: s/Takes/Calculates/

This is to ensure the same language as the package description.

```javascript
var Float32Array = require( '@stdlib/array/float32' );

var x = new Float32Array( [ 4.0, 2.0, -3.0, 5.0, -1.0 ] );
var y = new Float32Array( [ 2.0, 6.0, -1.0, -4.0, 8.0 ] );

> sdot( x.length, x, 1, y, 1 );

Reviewer: ```suggestion
var z = sdot( x.length, x, 1, y, 1 );
```

> // <number> -12.0

Reviewer: ```suggestion
// returns -12.0
```

```

The function accepts the following parameters:

-   **N**: number of indexed elements.
-   **x**: input [`Float32Array`][mdn-float32array].
-   **strideX**: index increment for `x`.
-   **y**: input [`Float32Array`][mdn-float32array].
-   **strideY**: index increment for `y`.

The `N` and `stride` parameters determine which elements in `x` and `y` are accessed at runtime. For example, to take the dot product of every other value in `x` and the first `N` elements of `y` in reverse order,

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

var x = new Float32Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
var y = new Float32Array( [ 1.0, 1.0, 1.0, 1.0, 1.0, 1.0 ] );

var N = floor( x.length / 2 );

> sdot( N, x, 2, y, -1 );

Reviewer: ```suggestion
var z = sdot( N, x, 2, y, -1 );
```

> // y => <number>9.0

Reviewer: ```suggestion
// returns 9.0
```

```

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

<!-- eslint-disable stdlib/capitalized-comments -->

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

// Initial arrays...
var x0 = new Float32Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
var y0 = new Float32Array( [ 7.0, 8.0, 9.0, 10.0, 11.0, 12.0 ] );

// Create offset views...
var x1 = new Float32Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
var y1 = new Float32Array( y0.buffer, y0.BYTES_PER_ELEMENT*3 ); // start at 4th element

var N = floor( x0.length / 2 );

> sdot( N, x1, -2, y1, 1 );

Reviewer: ```suggestion
var z = sdot( N, x1, -2, y1, 1 );
```

> // 92.0

Reviewer: ```suggestion
// returns 92.0
```

```

#### sdot.ndarray( N, x, strideX, offsetX, y, strideY, offsetY )

Calculates the dot product of `x` and `y`, with alternative indexing semantics.

```javascript
var Float32Array = require( '@stdlib/array/float32' );

var x = new Float32Array( [ 4.0, 2.0, -3.0, 5.0, -1.0 ] );
var y = new Float32Array( [ 2.0, 6.0, -1.0, -4.0, 8.0 ] );

> sdot.ndarray( x.length, x, 1, 0, y, 1, 0 );

Reviewer: ```suggestion
var z = sdot.ndarray( x.length, x, 1, 0, y, 1, 0 );
```

> // -12.0

Reviewer: ```suggestion
// returns -12.0
```

```

The function accepts the following additional parameters:

-   **offsetX**: starting index for `x`.
-   **offsetY**: starting index for `y`.

> While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offsetX` and `offsetY` parameters support indexing semantics based on starting indices. For example, to take the dot prodoct of every second value in `x` starting from the second value by the last 3 elements in `y` in reverse order.

Reviewer: ```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offsetX` and `offsetY` parameters support indexing semantics based on starting indices. For example, to take the dot product of every second value in `x` starting from the second value with the last 3 elements in `y` in reverse order
```

```javascript
var Float32Array = require( '@stdlib/array/float32' );
var floor = require( '@stdlib/math/base/special/floor' );

var x = new Float32Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
var y = new Float32Array( [ 7.0, 8.0, 9.0, 10.0, 11.0, 12.0 ] );

var N = floor( x.length / 2 );

> sdot.ndarray( N, x, 2, 1, y, -1, y.length-1 );

Reviewer: ```suggestion
var z = sdot.ndarray( N, x, 2, 1, y, -1, y.length-1 );
```

> // 120.0

Reviewer: ```suggestion
// returns 120.0
```

```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   If `N <= 0` both functions return `0.0`.

> -   `saxpy()` corresponds to the [BLAS][blas] level 1 function [`sdot`][sdot].

Reviewer: ```suggestion
-   `sdot()` corresponds to the [BLAS][blas] level 1 function [`sdot`][sdot].
```
